An error occurred when using the get_flops method of natten

I used the following code to calculate the model's parameter count:
  ```
from natten.flops import get_flops
    model = model.cuda(gpu)
    x = [torch.zeros(1, 3, 480, 640, dtype = torch.float32).cuda(gpu), torch.zeros(1, 3, 480, 640, dtype = torch.float32).cuda(gpu)]
    model.eval()
    flop_ctr = get_flops(model, x)
```
However, the following error occurred in the code:
```
line 101, in forward
    fuse_rgb = na2d(
  File "/data/miniconda3/envs/work1/lib/python3.10/site-packages/natten/functional.py", line 1734, in na2d
    tiling_config_forward, tiling_config_backward = autotune_fna(
  File "/data/miniconda3/envs/work1/lib/python3.10/site-packages/natten/autotuner/__init__.py", line 178, in autotune_fna
    best_backward_config = get_default_tiling_config_for_fna_backward(
  File "/data/miniconda3/envs/work1/lib/python3.10/site-packages/natten/autotuner/fna_backward.py", line 151, in get_default_tiling_config_for_fna_backward
    assert len(spatial_extent) == na_dim and all(
AssertionError
```

I checked the intermediate outputs of the function and found that the .shape return value in the code ` spatial_extent = tuple(x for x in input_tensor.shape[1 : na_dim + 1])` of the function `get_default_tiling_config_for_fna_backward` in is torch.int64 instead of the expected int.I don't know how the error occurred, as everything was fine when the model was being trained.

Thank you for your continued efforts to fix the issues; I really appreciate it! However, after trying the suggested branch and replacing all the .py files in my local NATTEN installation as instructed, I’m still encountering some problems.
```File "/data/miniconda3/envs/work1/lib/python3.10/site-packages/natten/flops.py", line 357, in get_flops
    return flop_ctr.total()
  File "/data/miniconda3/envs/work1/lib/python3.10/site-packages/fvcore/nn/jit_analysis.py", line 248, in total
    stats = self._analyze()
  File "/data/miniconda3/envs/work1/lib/python3.10/site-packages/fvcore/nn/jit_analysis.py", line 584, in _analyze
    op_counts = self._op_handles[kind](inputs, outputs)
  File "/data/miniconda3/envs/work1/lib/python3.10/site-packages/natten/flops.py", line 50, in fna_generic_flops
    len(output_shapes) == 2
AssertionError: Expected exactly 2 outputs (attention output, and LSE), got 1 ```

Thank you for your continued efforts to fix the issues; I really appreciate it! However, after trying the suggested branch and replacing all the .py files in my local NATTEN installation as instructed, I’m still encountering some problems.
```
File` "/data/miniconda3/envs/work1/lib/python3.10/site-packages/natten/flops.py", line 357, in get_flops
    return flop_ctr.total()
  File "/data/miniconda3/envs/work1/lib/python3.10/site-packages/fvcore/nn/jit_analysis.py", line 248, in total
    stats = self._analyze()
  File "/data/miniconda3/envs/work1/lib/python3.10/site-packages/fvcore/nn/jit_analysis.py", line 584, in _analyze
    op_counts = self._op_handles[kind](inputs, outputs)
  File "/data/miniconda3/envs/work1/lib/python3.10/site-packages/natten/flops.py", line 50, in fna_generic_flops
    len(output_shapes) == 2
AssertionError: Expected exactly 2 outputs (attention output, and LSE), got 1
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,9 @@
   compile.
   * In theory, any number of different attentions can be merged in this way, but the interface only
   allows one additional KV set for now.
+* Bug fixes in FLOP counter (with fvcore)
+  * Bugs introduced since FNA backwards were fixed.
+  * Finally added unit tests.
 
 ## [0.17.3] - 2024-11-01
 * Bug fix for torch < 2.4

--- a/requirements-dev.txt
+++ b/requirements-dev.txt
@@ -8,3 +8,4 @@ pytest==7.4.4
 click==8.1.7
 rich
 xformers>=0.25.0
+fvcore==0.1.5.post20221221

--- a/src/natten/autotuner/fna_backward.py
+++ b/src/natten/autotuner/fna_backward.py
@@ -145,7 +145,7 @@ def get_default_tiling_config_for_fna_backward(
 ) -> FnaBackwardConfigType:
     assert na_dim > 0 and na_dim < 4
     assert input_tensor.dim() == na_dim + 3
-    spatial_extent = tuple(x for x in input_tensor.shape[1 : na_dim + 1])
+    spatial_extent = tuple(int(x) for x in input_tensor.shape[1 : na_dim + 1])
     assert len(spatial_extent) == na_dim and all(
         isinstance(x, int) for x in spatial_extent
     )

--- a/src/natten/flops.py
+++ b/src/natten/flops.py
@@ -36,21 +36,43 @@ def fna_generic_flops(inputs: List[Any], outputs: List[Any]) -> Number:
         len(inputs) >= 3
     ), f"Expected at least 3 inputs (query, key, value), got {len(inputs)}"
     has_bias = len(inputs) == 4 and inputs[-1] is not None
-    assert len(outputs) >= 1, f"Expected at least 1 output, got {len(outputs)}"
     input_shapes = [get_shape(v) for v in inputs]
-    output_shapes = [get_shape(v) for v in outputs]
+
+    # Weird fvcore / jit bug
+    assert len(outputs) in [
+        1,
+        2,
+    ], f"Expected exactly 1 or 2 outputs (tuple), got {len(outputs)}"
+    if len(outputs) == 1:
+        outputs_ = outputs[0].uses()[0].user.outputs()
+        output_shapes = [get_shape(v) for v in outputs_]
+        assert (
+            len(output_shapes) == 2
+        ), f"Expected exactly 2 outputs (attention output, and LSE), got {len(output_shapes)}"
+    else:
+        assert (
+            len(outputs) == 2
+        ), f"Expected exactly 2 outputs (attention output, and LSE), got {len(outputs)}"
+        output_shapes = [get_shape(v) for v in outputs]
+
     assert len(input_shapes[0]) in [
         4,
         5,
         6,
     ], f"Input tensors must be of rank 4, 5, or 6, got {len(input_shapes[0])}"
+
     assert len(input_shapes[1]) == len(input_shapes[1]) == len(input_shapes[2]), (
         f"All input tensors must be of the same rank, got {len(input_shapes[0])}, "
         + f"{len(input_shapes[1])} and {len(input_shapes[2])}"
     )
+
     assert len(output_shapes[0]) == len(
         input_shapes[0]
     ), f"Output tensor must match the rank of input tensors, got {len(output_shapes[0])} != {len(input_shapes[0])}"
+    assert len(output_shapes[1]) == len(
+        output_shapes[0][:-1]
+    ), f"Mismatch between output and LSE shape, got {len(output_shapes[0])} != {len(output_shapes[1])}"
+
     assert input_shapes[0] == input_shapes[1] == input_shapes[2] == output_shapes[0], (
         "Query, key, value, and output must match in shape, got q.shape="
         + f"{input_shapes[0]}, k.shape={input_shapes[1]}, v.shape={input_shapes[1]}."
@@ -74,8 +96,24 @@ def fna_generic_flops(inputs: List[Any], outputs: List[Any]) -> Number:
         and callable(_uses[0].user.scalar_args)
     )
     scalar_args = _uses[0].user.scalar_args()
-    assert hasattr(scalar_args, "__len__") and len(scalar_args) == 5
-    kernel_size, dilation, is_causal, attn_scale, tiling_config = scalar_args
+    assert hasattr(scalar_args, "__len__") and len(scalar_args) in [
+        6,
+        7,
+    ], f"Expected 6 or 7 non-tensor args to the op, got {len(scalar_args)}. {scalar_args=}"
+
+    # Kick off rpb=None
+    if len(scalar_args) == 7:
+        assert scalar_args[0] is None
+        scalar_args = scalar_args[1:]
+
+    (
+        kernel_size,
+        dilation,
+        is_causal,
+        attn_scale,
+        tiling_config,
+        tiling_config_backward,
+    ) = scalar_args
     # TODO: it's very easy to hit this assertion. We must make sure
     # arguments like kernel size are checked before calling the autograd function,
     # not inside it.
@@ -129,9 +167,13 @@ def qk_1d_rpb_flop(inputs: List[Any], outputs: List[Any]) -> Number:
     ), f"Query and Key shapes did not match! Q: {input_shapes[0]}, K: {input_shapes[1]}"
     batch_size, heads, length, dim = input_shapes[0]
     batch_size, heads, length, kernel_size = output_shapes[0]
+    has_rpb = len(inputs) == 3
 
     flops = batch_size * heads * length * dim * kernel_size
+    # TODO: why is the sum of exps reduction not included in softmax's flops?
     flops += batch_size * heads * length * kernel_size
+    if has_rpb:
+        flops += batch_size * heads * length * kernel_size
     return flops
 
 
@@ -186,9 +228,13 @@ def qk_2d_rpb_flop(inputs: List[Any], outputs: List[Any]) -> Number:
     ), f"Query and Key shapes did not match! Q: {input_shapes[0]}, K: {input_shapes[1]}"
     batch_size, heads, height, width, dim = input_shapes[0]
     batch_size, heads, height, width, kernel_size_sq = output_shapes[0]
+    has_rpb = len(inputs) == 3
 
     flops = batch_size * heads * height * width * dim * kernel_size_sq
+    # TODO: why is the sum of exps reduction not included in softmax's flops?
     flops += batch_size * heads * height * width * kernel_size_sq
+    if has_rpb:
+        flops += batch_size * heads * height * width * kernel_size_sq
     return flops
 
 
@@ -243,9 +289,13 @@ def qk_3d_rpb_flop(inputs: List[Any], outputs: List[Any]) -> Number:
     ), f"Query and Key shapes did not match! Q: {input_shapes[0]}, K: {input_shapes[1]}"
     batch_size, heads, depth, height, width, dim = input_shapes[0]
     batch_size, heads, depth, height, width, kernel_size_cu = output_shapes[0]
+    has_rpb = len(inputs) == 3
 
     flops = batch_size * heads * depth * height * width * dim * kernel_size_cu
+    # TODO: why is the sum of exps reduction not included in softmax's flops?
     flops += batch_size * heads * depth * height * width * kernel_size_cu
+    if has_rpb:
+        flops += batch_size * heads * depth * height * width * kernel_size_cu
     return flops
 
 

--- a/src/natten/utils/testing.py
+++ b/src/natten/utils/testing.py
@@ -34,6 +34,13 @@ _HAS_GEMM_KERNELS = has_gemm()
 _GEMM_WITH_DOUBLE_PRECISION = has_fp64_gemm()
 _HAS_FNA_KERNELS = has_fna()
 
+try:
+    import fvcore  # type: ignore  # noqa: F401
+
+    _IS_FVCORE_AVAILABLE = True
+except ImportError:
+    _IS_FVCORE_AVAILABLE = False
+
 
 def skip_if_cuda_is_not_supported():
     def decorator(f):
@@ -100,6 +107,19 @@ def skip_if_nested_is_not_supported():
                 self.skipTest(
                     "Nested tensors are not supported with this torch version."
                 )
+            else:
+                return f(self, *args, **kwargs)
+
+        return wrapper
+
+    return decorator
+
+
+def skip_if_fvcore_is_not_available():
+    def decorator(f):
+        def wrapper(self, *args, **kwargs):
+            if not _IS_FVCORE_AVAILABLE:
+                self.skipTest("fvcore is not installed.")
             else:
                 return f(self, *args, **kwargs)
 

--- a/tests/test_flops.py
+++ b/tests/test_flops.py
@@ -39,7 +39,7 @@ def _reset_everything():
     natten.use_tiled_na()
     natten.use_gemm_na()
     natten.use_tf32_in_gemm_na()
-    torch.use_deterministic_algorithms(False)
+    natten.use_fused_na(False, kv_parallel=False)
     os.environ["NATTEN_LOG_LEVEL"] = "CRITICAL"
 
 

--- a/tests/test_flops.py
+++ b/tests/test_flops.py
@@ -1,0 +1,292 @@
+#################################################################################################
+# Copyright (c) 2022-2024 Ali Hassani.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+#################################################################################################
+
+import math
+import os
+import unittest
+
+import natten
+
+import torch
+from natten.utils.testing import (
+    skip_if_cuda_is_not_supported,
+    skip_if_fna_is_not_supported,
+    skip_if_fvcore_is_not_available,
+)
+
+
+def _reset_everything():
+    natten.use_tiled_na()
+    natten.use_gemm_na()
+    natten.use_tf32_in_gemm_na()
+    torch.use_deterministic_algorithms(False)
+    os.environ["NATTEN_LOG_LEVEL"] = "CRITICAL"
+
+
+class FlopCounterTests(unittest.TestCase):
+    def setUp(self):
+        _reset_everything()
+
+    def tearDown(self):
+        _reset_everything()
+
+    def _build_model(
+        self,
+        dim_per_head,
+        heads,
+        kernel_size,
+        dilation,
+        is_causal,
+        rel_pos_bias,
+        qkv_bias,
+    ):
+        mod = natten.NeighborhoodAttention1D
+        if hasattr(kernel_size, "__len__") and len(kernel_size) == 2:
+            mod = natten.NeighborhoodAttention2D
+        elif hasattr(kernel_size, "__len__") and len(kernel_size) == 3:
+            mod = natten.NeighborhoodAttention3D
+
+        return mod(
+            dim=dim_per_head * heads,
+            num_heads=heads,
+            kernel_size=kernel_size,
+            dilation=dilation,
+            is_causal=is_causal,
+            rel_pos_bias=rel_pos_bias,
+            qkv_bias=qkv_bias,
+        )
+
+    def _build_input(self, batch, spatial_extent, heads, dim_per_head):
+        shape = [batch, *spatial_extent, heads * dim_per_head]
+        return torch.randn(shape)
+
+    def _compute_flops(
+        self,
+        batch,
+        spatial_extent,
+        heads,
+        dim_per_head,
+        kernel_size,
+        dilation,
+        is_causal,
+        rel_pos_bias,
+        qkv_bias,
+    ):
+        # TODO: why is FVCore reporting MACs, not FLOPs?
+        # c = 2
+        c = 1
+
+        qkv_M, qkv_N, qkv_K = (
+            batch * math.prod(spatial_extent),
+            dim_per_head * heads * 3,
+            dim_per_head * heads,
+        )
+        qkv_flops = qkv_M * qkv_N * qkv_K * c
+        # TODO: FVCore doesn't count bias FLOPs
+        # if qkv_bias:
+        #     qkv_flops += qkv_M * qkv_N
+
+        proj_M, proj_N, proj_K = (
+            batch * math.prod(spatial_extent),
+            dim_per_head * heads,
+            dim_per_head * heads,
+        )
+        proj_flops = proj_M * proj_N * proj_K * c
+
+        attn_0_M, attn_0_N, attn_0_K = (
+            batch * heads * math.prod(spatial_extent),
+            math.prod(kernel_size),
+            dim_per_head,
+        )
+        attn_1_M, attn_1_N, attn_1_K = (
+            batch * heads * math.prod(spatial_extent),
+            dim_per_head,
+            math.prod(kernel_size),
+        )
+
+        attn_0_flops = attn_0_M * attn_0_N * attn_0_K * c
+        attn_1_flops = attn_1_M * attn_1_N * attn_1_K * c
+
+        # TODO: why is the sum of exps reduction not included in softmax's flops?
+        softmax_M, softmax_K = batch * heads * math.prod(spatial_extent), math.prod(
+            kernel_size
+        )
+        softmax_flops = softmax_M * softmax_K
+
+        attn_flops = attn_0_flops + attn_1_flops + softmax_flops
+
+        if rel_pos_bias:
+            rpb_M, rpb_K = batch * heads * math.prod(spatial_extent), math.prod(
+                kernel_size
+            )
+            attn_flops += rpb_M * rpb_K
+
+        return qkv_flops + attn_flops + proj_flops
+
+    def _test_natten_flops(
+        self,
+        batch,
+        spatial_extent,
+        heads,
+        dim_per_head,
+        kernel_size,
+        dilation,
+        is_causal,
+        rel_pos_bias,
+        qkv_bias,
+        run_on_cuda=False,
+    ):
+        try:
+            from natten.flops import get_flops
+        except ImportError:
+            raise ImportError(
+                "FVCore not found, but related tests are still being run. "
+                "Did you forget to use @skip_if_fvcore_is_not_available ?"
+            )
+        m = self._build_model(
+            dim_per_head=dim_per_head,
+            heads=heads,
+            kernel_size=kernel_size,
+            dilation=dilation,
+            is_causal=is_causal,
+            rel_pos_bias=rel_pos_bias,
+            qkv_bias=qkv_bias,
+        )
+        x = self._build_input(
+            batch=batch,
+            spatial_extent=spatial_extent,
+            heads=heads,
+            dim_per_head=dim_per_head,
+        )
+        if run_on_cuda:
+            m = m.cuda()
+            x = x.cuda()
+
+        natten_flops = get_flops(m, x)
+
+        reference_flops = self._compute_flops(
+            batch=batch,
+            spatial_extent=spatial_extent,
+            heads=heads,
+            dim_per_head=dim_per_head,
+            kernel_size=kernel_size,
+            dilation=dilation,
+            is_causal=is_causal,
+            rel_pos_bias=rel_pos_bias,
+            qkv_bias=qkv_bias,
+        )
+
+        assert reference_flops == natten_flops, (
+            "FLOPs are incorrect. "
+            + f"Expected {reference_flops} FLOPs, FVCore computed {natten_flops} FLOPs. "
+            + f"Difference: {abs(reference_flops - natten_flops)} FLOPs. "
+        )
+
+    @skip_if_fvcore_is_not_available()
+    def test_flops_unfused(self):
+        natten.use_fused_na(False, kv_parallel=False)
+
+        self._test_natten_flops(
+            batch=4,
+            spatial_extent=(128,),
+            heads=4,
+            dim_per_head=64,
+            kernel_size=(5,),
+            dilation=(2,),
+            is_causal=(False,),
+            rel_pos_bias=False,
+            qkv_bias=False,
+        )
+
+        self._test_natten_flops(
+            batch=1,
+            spatial_extent=(16, 10),
+            heads=2,
+            dim_per_head=32,
+            kernel_size=(5, 3),
+            dilation=(2, 1),
+            is_causal=(False, False),
+            rel_pos_bias=True,
+            qkv_bias=True,
+        )
+
+        self._test_natten_flops(
+            batch=1,
+            spatial_extent=(6, 5, 8),
+            heads=2,
+            dim_per_head=32,
+            kernel_size=(5, 3, 7),
+            dilation=(1, 1, 1),
+            is_causal=(True, False, False),
+            rel_pos_bias=False,
+            qkv_bias=True,
+        )
+
+    @skip_if_fvcore_is_not_available()
+    @skip_if_cuda_is_not_supported()
+    @skip_if_fna_is_not_supported()
+    def test_flops_fna(self):
+        natten.use_fused_na(True, kv_parallel=True)
+
+        self._test_natten_flops(
+            batch=4,
+            spatial_extent=(128,),
+            heads=4,
+            dim_per_head=64,
+            kernel_size=(5,),
+            dilation=(2,),
+            is_causal=(False,),
+            rel_pos_bias=False,
+            qkv_bias=False,
+            run_on_cuda=True,
+        )
+
+        self._test_natten_flops(
+            batch=1,
+            spatial_extent=(16, 10),
+            heads=2,
+            dim_per_head=32,
+            kernel_size=(5, 3),
+            dilation=(2, 1),
+            is_causal=(False, False),
+            rel_pos_bias=True,
+            qkv_bias=True,
+            run_on_cuda=True,
+        )
+
+        self._test_natten_flops(
+            batch=1,
+            spatial_extent=(6, 5, 8),
+            heads=2,
+            dim_per_head=32,
+            kernel_size=(5, 3, 7),
+            dilation=(1, 1, 1),
+            is_causal=(True, False, False),
+            rel_pos_bias=False,
+            qkv_bias=True,
+            run_on_cuda=True,
+        )
+
+
+if __name__ == "__main__":
+    unittest.main()

